### ISSUE TITLE
BBC: Blead Breaks Devel::Cover

### ISSUE THREAD

**Issue by cjg-cguevara** [Perl/perl5#19869]
This is a bug report for perl from "Carlos Guevara" <carlos@carlosguevara.com>,
generated with the help of perlbug 1.42 running under perl 5.37.1.

-----------------------------------------------------------------
[Please describe your issue here]

BBC: Blead Breaks Devel::Cover

Please see http://matrix.cpantesters.org/?dist=Devel::Cover

[Please do not change anything below this line]
-----------------------------------------------------------------
---
Flags:
    category=core
    severity=low
---
Site configuration information for perl 5.37.1:

Configured by cpan at Wed Jun 15 11:38:42 EDT 2022.

Summary of my perl5 (revision 5 version 37 subversion 1) configuration:
  Commit id: 0bf91cc9654c61052c0c525b894604f08b8d3069
  Platform:
    osname=openbsd
    osvers=7.1
    archname=OpenBSD.amd64-openbsd-thread-multi
    uname='openbsd cjg-openbsd7 7.1 generic#3 amd64 '
    config_args='-des -Dprefix=~/bin/perl-blead -Dscriptdir=~/bin/perl-blead/bin -Dusedevel -Duse64bitall -Duseithreads'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-pthread -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    optimize='-O2'
    cppflags='-pthread -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='OpenBSD Clang 13.0.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags ='-pthread -Wl,-E  -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/lib/clang/13.0.0/lib /usr/lib /usr/local/lib
    libs=-lpthread -lm -lutil -lc
    perllibs=-lpthread -lm -lutil -lc
    libc=/usr/lib/libc.so.96.1
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags='-DPIC -fPIC '
    lddlflags='-shared -fPIC  -L/usr/local/lib -fstack-protector-strong'

---
@INC for perl 5.37.1:
    /home/cpan/bin/perl-blead/lib/site_perl/5.37.1/OpenBSD.amd64-openbsd-thread-multi
    /home/cpan/bin/perl-blead/lib/site_perl/5.37.1
    /home/cpan/bin/perl-blead/lib/5.37.1/OpenBSD.amd64-openbsd-thread-multi
    /home/cpan/bin/perl-blead/lib/5.37.1

---
Environment for perl 5.37.1:
    HOME=/home/cpan
    LANG (unset)
    LANGUAGE (unset)
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/cpan/bin/perl-blead/bin:/home/cpan/bin:/home/cpan/bin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11R6/bin:/usr/local/bin:/usr/local/sbin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

**Comment by hvds** [comment]
I note that it fails in `t/e2e/aoverload_bool.t`, which may be a clue; I took a quick look, but the test process looks pretty opaque.

The build warns:
```
Devel::Cover 1.38 has not been tested with perl 5.037001.
Testing will take place against expected output from perl 5.036.
You may well find failing tests.
```

So I think we should leave it to @pjcj for initial investigation.

**Comment by jkeenan** [comment]
In our earlier discussion of this BBC issue in Devel-Cover, we didn't attempt to identify the problematic commit.  Today I got this:
```
c1bee3910fc96b0524cf78054c1bb35d8b5a322a is the first bad commit
commit c1bee3910fc96b0524cf78054c1bb35d8b5a322a
Author: Karl Williamson <khw@cpan.org>
Date:   Mon Jun 29 09:16:00 2020 -0600
Commit:     Karl Williamson <khw@cpan.org>
CommitDate: Mon Jun 13 06:05:09 2022 -0600


    Evaluate SvSetSV_(nosteal_)?and() params just once

 sv.h | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)
bisect run success
That took 2428 seconds.
```
@khwilliamson, can you take a look?

**Comment by khwilliamson** [comment]
On 7/10/22 16:15, James E Keenan wrote:
> In our earlier discussion of this BBC issue in Devel-Cover, we didn't 
> attempt to identify the problematic commit. Today I got this:
> 
> |c1bee3910fc96b0524cf78054c1bb35d8b5a322a is the first bad commit commit 
> c1bee3910fc96b0524cf78054c1bb35d8b5a322a Author: Karl Williamson 
> ***@***.***> Date: Mon Jun 29 09:16:00 2020 -0600 Commit: Karl 
> Williamson ***@***.***> CommitDate: Mon Jun 13 06:05:09 2022 -0600 
> Evaluate SvSetSV_(nosteal_)?and() params just once sv.h | 16 
> ++++++++++++---- 1 file changed, 12 insertions(+), 4 deletions(-) bisect 
> run success That took 2428 seconds. |
> 
> @khwilliamson <https://github.com/khwilliamson>, can you take a look?
> 

I looked at this, and found the test process opaque as well.  But what I 
think is happening is that the overloaded function 'meh' is getting 
called one less time than before.  This could easily be the result of 
the blamed commit, which changes to evaluate its argument just once, and 
the test suite has hard-coded in an expected extra call.

**Comment by haarg** [comment]
I've created a PR that should fix this.

pjcj/Devel--Cover#304

**Comment by jkeenan** [comment]
On 6/15/22 21:39, Carlos Guevara wrote:
> This is a bug report for perl from "Carlos Guevara" 
> ***@***.*** ***@***.***>,
> generated with the help of perlbug 1.42 running under perl 5.37.1.
> 
> ------------------------------------------------------------------------
> 
> [Please describe your issue here]
> 
> BBC: Blead Breaks Devel::Cover
> 
> Please see http://matrix.cpantesters.org/?dist=Devel::Cover 
> <http://matrix.cpantesters.org/?dist=Devel::Cover>
> 

Devel::Cover is highly sensitive to changes in the Perl internals and so 
breaks at least once per production release.  Paul Johnson will take 
care of this at some point during the cycle.  So we simply need to put a 
BBC label on this issue and not worry about it for several months.

**Comment by pjcj** [comment]
Thanks everyone for the work put into this!  I have released Devel::Cover 1.39 which fixes this and passes tests which were failing due to other changes in the 5.37 cycle.  I still need to track down some of the changes and make fixes in Devel::Cover, which I will do in subsequent releases, but for now B doesn't B Devel::Cover so please feel free to close this ticket.
